src/Client/HedgedConnections.cpp

```diff
@@ -296,7 +296,7 @@ HedgedConnections::ReplicaLocation HedgedConnections::getReadyReplicaLocation(As
     {
         ReplicaLocation location = replica_with_last_received_packet.value();
         replica_with_last_received_packet.reset();
-        if (resumePacketReceiver(location))
+        if (offset_states[location.offset].replicas[location.index].connection->hasReadPendingData() && resumePacketReceiver(location))
```

Comment:
But why do we need to check `offset_states[location.offset].replicas[location.index].connection->hasReadPendingData()` here?
I suppose, even if connection does not have any data, we just need to receive bytes from socket synchronously (and add socket fd to epoll and so on ...)

Reply:
Ok, we don't remove socket from epoll.
But looks like we need to initiate reading anyway.
I guess it works because server send progress periodically.

Reply:
> But looks like we need to initiate reading anyway.
Isn't it more logical to initiate reading when there is data to read (socket is ready or pending data in buffer)? If there is no data we will do some unnecessary actions (resume fiber, set callback, try to read and call async_callback). For example, when there is no data in any socket and we call `receivePacketUnlocked(AsyncCallback async_callback)` I guess we expect that we will call `async_callback` immediately. I made some tests, without `hasReadPendingData()` check async_remote_read performance test shows a worse result than with this check.